### PR TITLE
:sparkles: Add course setting menu for system staff at preference　#50

### PR DESCRIPTION
**Pull request by j-nakashima**
I'd like to be received a review of my code .

I have one concern.
- Manager's name in search condition works differently, if two words are inputted or not.
  If two words separated with blank are inputted,  first word is regarded as family name and second word as given name.
 Otherwise, word is regarded as both family name and given name.
 There are two cases in manager's search I think.
 One is to search with full name, and the other is to search with family name or given name.
 Is it misleading?
(ex) 
  "Oita Taro" -> search with "Oita" as family name and "Taro" as given name.
  "Oita"  -> search with "Oita" as family name or given name.

### CODE REVIEW

File: app/models/course.rb [hunk 1]
```diff
@@ -265,9 +265,34 @@ def eval_category
     end
   end
 
+  def self.search(term_id, status, title_parts, manager_parts)
+    @candidates = Course.all.limit(COURSE_SEARCH_MAX_SIZE)
```
Comment (kyoshizaki) [inline app/models/course.rb]: In this stage, it is better to use only,

`@candidates = Course.all`

File: app/models/course.rb [hunk 2]
```diff
@@ -265,9 +265,34 @@ def eval_category
...
+        logical_operation = 'and'
+      else
+        family_name = given_name = manager_parts
+        logical_operation = 'or'
+      end
+      manager_ids = User.where('family_name like ? ' + logical_operation + ' given_name like ? ', '%' + family_name + '%', '%' + given_name + '%').pluck(:id)
+      course_ids = CourseMember.where(role: 'manager').where('user_id IN (?)', manager_ids).pluck(:course_id).uniq
+      @candidates = @candidates.where('id IN(?)', course_ids)
+    end
+    @candidates
```
Comment (kyoshizaki) [inline app/models/course.rb]: Here, limit the candidates number.

`@candidates.limit(COURSE_SEARCH_MAX_SIZE)`

File: app/models/course.rb [hunk 3]
```diff
@@ -265,9 +265,34 @@ def eval_category
     end
   end
 
+  def self.search(term_id, status, title_parts, manager_parts)
+    @candidates = Course.all.limit(COURSE_SEARCH_MAX_SIZE)
+    @candidates = @candidates.where(term_id: term_id) unless term_id.empty?
+    @candidates = @candidates.where(status: status) unless status.empty?
+    @candidates = @candidates.where('title like ?', '%' + title_parts + '%') if title_parts.present?
+    if manager_parts.present?
```
Comment (kyoshizaki) [inline app/models/course.rb]: See PR comment.

File: config/locales/translation_en.yml [hunk 4]
```diff
@@ -99,6 +99,9 @@ en:
         overview: Overview
         staff_course_notes: Staff course notes
         status: Status
+        status_archived: Archived
```
Comment (kyoshizaki) [inline config/locales/translation_en.yml]: See comment in translation_ja.yml

File: config/locales/translation_ja.yml [hunk 5]
```diff
@@ -98,6 +98,9 @@ ja:
         overview: 概要
         staff_course_notes: スタッフのコースノート
         status: 状態
+        status_archived: アーカイブ
```
Comment (kyoshizaki) [inline config/locales/translation_ja.yml]: For a while, translations for attribute values (not attribute names themselves) are better to locate under activrecords.others.
So in this case, put these status_* under activrecords.others.course.status.* in YAML file.

See also comment in issue #52

[the remark this answers is not in the document]
Comment (j-nakashima) [reply config/locales/translation_ja.yml]: :ok_woman:

Comment (j-nakashima) [reply config/locales/translation_ja.yml]: :ok_woman:

File: config/locales/translation_en.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (j-nakashima) [reply config/locales/translation_en.yml]: :ok_woman:

Comment (j-nakashima) [reply config/locales/translation_en.yml]: :ok_woman:

File: app/views/courses/_course_candidates.html.erb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        </td>
+        <td class="name-td">
+          <%= course.term.title %>
+        </td>
+        <% managers = course_managers_display_list(course) %>
+        <td class="name-td">
+          <%= course.title  %> 
+        </td>
+        <td class="name-td">
+           <a href="#" class="Label label-info" data-toggle="tooltip" data-original-title="<%= managers %>">
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_course_candidates.html.erb]: :ok: I got it.

File: app/views/courses/_course_candidates.html.erb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            <%= managers.truncate(COURSE_MANAGER_DISPLAY_MAX_LENGTH) %></a>
+        </td>
+        <td class="name-td">
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_course_candidates.html.erb]: :ok:

File: app/controllers/courses_controller.rb [hunk 8]
```diff
@@ -270,12 +270,45 @@ def ajax_sort_lessons
     render 'layouts/renders/resource', locals: { resource: 'courses/edit_lessons' }
   end
 
+  def ajax_course_pref
+    # for security reason
+    if User.system_staff? session[:id]
+      @user = User.find session[:id]
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/controllers/courses_controller.rb]: I got it.

File: app/controllers/courses_controller.rb [hunk 9]
```diff
@@ -270,12 +270,45 @@ def ajax_sort_lessons
...
+
+  def ajax_search_courses
+    @candidates = Course.search(params[:term_id], params[:status], params[:title], params[:manager])
+    if @candidates.size.zero? || (!User.system_staff? session[:id])
+      ajax_index_no_course
+    end
+    render 'layouts/renders/main_pane_candidates', locals: { resource: 'select_course' }
+  end
+
+  def ajax_index_by_manager
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/controllers/courses_controller.rb]: Change to "ajax_index_by_system_staff".

File: app/models/course.rb (hunk 2 above)
Comment (j-nakashima) [reply app/models/course.rb]: :question: May I ask a question?
The  sql I executed searching was as follows in the current source code.
-- 
SELECT  "courses".* FROM "courses" WHERE "courses"."term_id" = ? AND "courses"."status" = ? LIMIT ?  [["term_id", 1], ["status", "open"], ["LIMIT", 50]]
--
It seems "Rails" automatically assemble the efficient sql.
Is this depend on the environment?

File: app/views/courses/_search_courses.html.erb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+	<h2>1. <%= t('.search_course') %>
+		<div class="header-explanation"><%= t('.exp_search_course', num: COURSE_SEARCH_MAX_SIZE) %></div>
+	</h2>
+	<div style="margin:2px">
+		<%= form_tag({action: 'ajax_search_courses'}, class: 'form-group', remote: true) do %>
+		<div class="form-group">
+			<div class="form-group form-inline">
+				<div class="form-group col-md-5">
+			<%# --- term --- %>
+			<%= label :term_id, t('activerecord.models.term') + ':' , class: 'col-md-3 col-form-label text-sm-right' %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :ok: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+			<%= select_tag :term_id, options_for_select(select_options_unspecified(Term.selectables("all"))), class: 'form-control input-lg' %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :ok: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+				</div>
+				<div class="form-group col-md-7">
+			<%# --- title --- %>
+			<%= label :title, t('activerecord.attributes.course.title') + ':' , class: 'col-md-3 col-form-label text-sm-right' %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+			<%= text_field_tag :title, @search_word, class: 'form-control input-lg', placeholder: t('.placeholder_title') %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+				</div>
+			</div>
+			<div class="form-group form-inline">
+				<div class="form-group col-md-5">
+			<%# --- status --- %>
+			<%= label :status, t('activerecord.attributes.course.status') + ':' , class: 'col-md-3 col-form-label text-sm-right' %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+			<%= select_tag :status, options_for_select(select_options_unspecified(status_select_options)), class: 'form-control input-lg' %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+				</div>
+				<div class="form-group col-md-7">
+			<%# --- manager --- %>
+			<%= label :manager, t('activerecord.attributes.course.managers') + ':', class: 'col-md-3 col-form-label text-sm-right'  %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/views/courses/_search_courses.html.erb [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+			<%= text_field_tag :manager, @search_word, class: 'form-control input-lg', placeholder:  t('.placeholder_manager') %>
```
[the remark this answers is not in the document]
Comment (j-nakashima) [reply app/views/courses/_search_courses.html.erb]: :wrench: I got it.

File: app/models/course.rb (hunk 2 above)
Comment (kyoshizaki) [reply app/models/course.rb]: This was my misunderstanding.
The active record query methods are executed for the first time when those are needed.
So in this case, the same result will be obtained regardless of the execution place of the limit methods.

But I think it is easier to understand if limit is listed after where.

**Pull request merged**